### PR TITLE
Merge Stitch and Realm SDKs - 5: remote collection, added iterables and preliminary support for find and aggregate

### PR DESCRIPTION
**Pull request by edualonso**
Merge Stitch and Realm SDKs - 5:

- Preliminary support for find and aggregate operations requires adding and adapting some iterables from Stitch.
- Big cleanup and code simplification in OsMongoClient Java and C++ classes.
- Added more test scenarios now that the Bson parser works.

There are still two pending issues which block the full validation of this task:

- https://github.com/realm/realm-object-store/pull/1013 - fixed and merged
- https://jira.mongodb.org/browse/REALMC-5787 - no estimate on this one

### CODE REVIEW

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java [hunk 1]
```diff
@@ -406,61 +467,51 @@ protected ResultT mapSuccess(Object result) {
...
+            case FIND_ONE_AND_REPLACE:
+                nativeFindOneAndReplace(type, nativePtr, encodedFilter, encodedUpdate, encodedProjection, encodedSort, false, false,callback);
                 break;
-            case REPLACE:
+            case FIND_ONE_AND_REPLACE_WITH_OPTIONS:
                 if (options == null) {
-                    nativeFindOneAndReplace(nativePtr, encodedFilter, encodedUpdate, callback);
-                } else {
-                    nativeFindOneAndReplaceWithOptions(nativePtr, encodedFilter, encodedUpdate, encodedProjection, encodedSort, options.isUpsert(), options.isReturnNewDocument(), callback);
+                    throw new IllegalStateException("FindOneAndModifyOptions must not be null");
```
Comment (rorbech) [inline realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java:502]: Shouldn't this be `IllegalArgumentException`? Here and other places

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java [hunk 2]
```diff
@@ -73,14 +90,18 @@ public long getNativeFinalizerPtr() {
...
     }
 
-    public Long count(@Nullable final Bson filter) {
-        return count(filter, null);
+    public Long count(final Bson filter, final CountOptions options) {
+        return countInternal(filter, options);
     }
 
-    public Long count(@Nullable final Bson filter, @Nullable final CountOptions options) {
+    private Long countInternal(final Bson filter, @Nullable final CountOptions options) {
```
Comment (rorbech) [inline realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java:104]: Did you intentionally remove the public possibility to call count with `options` without `filter`?

File: realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsMongoCollection.cpp [hunk 3]
```diff
@@ -116,41 +116,36 @@ Java_io_realm_internal_objectstore_OsMongoCollection_nativeCount(JNIEnv* env,
...
+                bson::BsonDocument sort(JniBsonProtocol::parse_checked(env, j_sort, Bson::Type::Document, "BSON sort must be a Document"));
+                RemoteMongoCollection::RemoteFindOptions options = {
+                        limit,
+                        projection,
+                        sort
+                };
+
+                collection->find_one(filter, options, JavaNetworkTransport::create_result_callback(env, j_callback, collection_mapper_find_one));
+                break;
+        }
```
Comment (rorbech) [inline realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsMongoCollection.cpp:148]: Maybe throw an error if it is an unexpected type, as you do in some of the other similar methods.

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.realm.ObjectServerError;
+import io.realm.internal.jni.JniBsonProtocol;
+import io.realm.internal.jni.OsJNIResultCallback;
+import io.realm.internal.network.ResultHandler;
+
+public abstract class OsMongoIterable<ResultT> {
```
Comment (rorbech) [inline realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java:30]: Seems to remember some comment about only using `Os`-prefix if it actually wrapped an underlying OS object.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt [hunk 5]
```diff
@@ -389,66 +363,151 @@ class MongoCollectionTest {
...
+            insertMany(listOf(doc1, doc2)).blockingGetResult()
+
+            assertTrue(iter.iterator().blockingGetResult()!!
+                    .hasNext().blockingGetResult()!!)
+            assertEquals(doc1.withoutId(), iter.first().blockingGetResult()!!.withoutId())
+            assertEquals(doc2.withoutId(),
+                    iter.limit(1)
+                            .sort(Document("_id", -1))
+                            .iterator().blockingGetResult()!!
+                            .next().blockingGetResult()!!
```
Comment (rorbech) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt:455]: Does this _double asynchronicity_ make sense? Doesn't it only involve one remote call?

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt [hunk 6]
```diff
@@ -228,37 +224,26 @@ class MongoCollectionTest {
...
+                    .sort(Document("hello", 1))
+            assertEquals(doc1, findOne(Document(), options1).blockingGetResult()!!.withoutId())
+
+            val options2 = FindOptions()
+                    .limit(2)
+                    .projection(projection)
+                    .sort(Document("hello", -1))
+            assertEquals(doc3.withoutId(), findOne(Document(), options2).blockingGetResult()!!.withoutId())
+
+            assertFails { findOne(Document("\$who", 1)).blockingGetResult() }
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt:246]: Yes, that exception handling is part of my next commit.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoIterable.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+     * @return first entry in the collection.
+     */
+    abstract ResultT getFirst();
+
+    /**
+     * Returns a cursor of the operation represented by this iterable.
+     *
+     * @return a cursor of the operation represented by this iterable.
+     */
+    public Task<MongoCursor<ResultT>> iterator() {
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoIterable.java:55]: The Javadoc might not be the best. I should probably rewrite it. This class is not a Java `Iterable`, although it tries to mimic its behaviour. The results need to be delivered as `Task`s though. This also affects the `MongoCursor` class, whose implementation I've changed for the next commit. It didn't need the second level of asynchronicity.

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java (hunk 4 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java:30]: I talked about it with @cmelchior and it's OK to use it for helper classes which don't instantiate objects - unless I misunderstood it @cmelchior?

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt (hunk 5 above)
Comment (edualonso) [reply realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt:455]: That's right. I have changed it for the next commit - see https://github.com/realm/realm-java/pull/6881/files#r430916108

File: realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsMongoCollection.cpp (hunk 3 above)
Comment (edualonso) [reply realm/realm-library/src/main/cpp/io_realm_internal_objectstore_OsMongoCollection.cpp:148]: Yes, totally forgot about it.

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java (hunk 2 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java:104]: It was a mistake, it never was part of the original API.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt [hunk 8]
```diff
@@ -51,6 +54,9 @@ class MongoCollectionTest {
     @Before
     fun setUp() {
         Realm.init(InstrumentationRegistry.getInstrumentation().targetContext)
+
+        RealmLog.setLevel(LogLevel.DEBUG)
```
Comment (edualonso) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt:58]: I will remove this, for it stays for now until all the OS bugs are fixed.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt [hunk 9]
```diff
@@ -51,6 +54,9 @@ class MongoCollectionTest {
...
+
```
Comment (edualonso) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoCollectionTest.kt:59]: I will also break down some of the tests into smaller, more meaningful cases in the upcoming commits.

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java [hunk 10]
```diff
@@ -90,99 +111,142 @@ protected Long mapSuccess(Object result) {
...
 
         nativeCount(nativePtr, filterString, limit, callback);
 
         return ResultHandler.handleResult(success, error);
     }
 
-    public Collection<DocumentT> find() {
-        return find(new Document());
-    }
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (edualonso) [inline realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java:122]: I will remove these separators. I added them for temporary convenience to quickly see the different operation blocks.

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java [hunk 11]
```diff
@@ -406,61 +467,51 @@ protected ResultT mapSuccess(Object result) {
 
         String encodedFilter = JniBsonProtocol.encode(filter, codecRegistry);
         String encodedUpdate = JniBsonProtocol.encode(update, codecRegistry);
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java]: Doesn't apply any longer.

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java [hunk 12]
```diff
@@ -221,42 +306,114 @@ protected InsertManyResult mapSuccess(Object result) {
...
+        String jsonUpdate = JniBsonProtocol.encode(update, codecRegistry);
+
+        switch (type) {
+            case UPDATE_ONE:
+            case UPDATE_MANY:
+                nativeUpdate(type, nativePtr, jsonFilter, jsonUpdate, false, callback);
+                break;
+            case UPDATE_ONE_WITH_OPTIONS:
+            case UPDATE_MANY_WITH_OPTIONS:
+                if (options == null) {
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoCollection.java:396]: We have `Util.checkNull(val, args)` for this kind of common checks

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java (hunk 4 above)
Comment (cmelchior) [reply realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java:31]: I think there was a bit of a misunderstanding then. Normally we only use `Os<X>` for classes wrapping some ObjectStore class. I believe we chatted about finalizers? 🤔 

So, in this case, it should probably be merged with the public MongoIterable

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java [hunk 13]
```diff
@@ -62,7 +66,30 @@
...
-        throw new RuntimeException("Not Implemented");
+        return nameSpace;
+    }
+
+    /**
+     * Get the class of documents stored in this collection.
+     * <p>
+     * If you used the simple {@link MongoDatabase#getCollection(String)} to get
+     * this collection,
+     * this is {@link org.bson.Document}.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java:77]: Wrapping looks wrong here

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java [hunk 14]
```diff
@@ -62,7 +66,30 @@
...
+     * </p>
+     *
+     * @return the class
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java:80]: This description doesn't say much. Should probably be expanded a bit.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java [hunk 15]
```diff
@@ -62,7 +66,30 @@
...
+     */
+    public Class<DocumentT> getDocumentClass() {
+        return osMongoCollection.getDocumentClass();
+    }
+
+    /**
+     * Get the codec registry for the RemoteMongoCollection.
+     *
+     * @return the {@link CodecRegistry}
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java:89]: This kind of JavaDoc doesn't really say anything the types dont already. I would probably rephrase to `@return the {@link CodecRegistry}  for this collection`.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java [hunk 16]
```diff
@@ -212,18 +234,18 @@ public MongoNamespace getNamespace() {
...
-    // FIXME: fix find implementation - ignore this code for code review
-    public Task<Iterable<DocumentT>> find(final FindOptions options) {
-        return dispatcher.dispatchTask(() ->
-                osMongoCollection.find(options)
-        );
+    /**
+     * Finds all documents in the collection that match the given filter.
+     *
+     * @param options a {@link FindOptions} struct
+     * @return the find iterable interface
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java:245]: Not sure what this means?

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java [hunk 17]
```diff
@@ -233,18 +255,21 @@ public MongoNamespace getNamespace() {
...
-        return dispatcher.dispatchTask(() ->
-                osMongoCollection.find(resultClass, options)
-        );
+    /**
+     * Finds all documents in the collection that match the given filter.
+     *
+     * @param resultClass the class to decode each document into
+     * @param options a {@link FindOptions} struct
+     * @param <ResultT>   the target document type of the iterable.
+     * @return the find iterable interface
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java:268]: Same for this and most other JavaDoc....this description doesn't say much.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+    private OsAggregateIterable<ResultT> osAggregateIterable;
+
+    public AggregateIterable(final TaskDispatcher dispatcher,
+                             final OsAggregateIterable<ResultT> osAggregateIterable) {
+        super(dispatcher);
+        this.osAggregateIterable = osAggregateIterable;
+    }
+
+    @Override
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java:41]: Missing Javadoc

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+    Collection<ResultT> getCollection() {
+        return osAggregateIterable.getCollection();
+    }
+
+    @Override
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java:46]: Same

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/FindIterable.java [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+    private final OsFindIterable<ResultT> osFindIterable;
+
+    public FindIterable(final TaskDispatcher dispatcher,
+                        final OsFindIterable<ResultT> osFindIterable) {
+        super(dispatcher);
+        this.osFindIterable = osFindIterable;
+    }
+
+    @Override
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/FindIterable.java:44]: Javadoc

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/FindIterable.java [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    Collection<ResultT> getCollection() {
+        return osFindIterable.getCollection();
+    }
+
+    @Override
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/FindIterable.java:49]: Javadoc

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoCursor.java [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+package io.realm.mongodb.mongo.iterable;
+
+import com.google.android.gms.tasks.Task;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * The Mongo Cursor class.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoCursor.java:25]: This doesn't say anything the name of the class doesn't. Maybe expand on what the class actually is for?

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoCursor.java [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    @Override
+    public boolean hasNext() {
+        return iterator.hasNext();
+    }
+
+    /**
+     * Returns the next document.
+     *
+     * @return A {@link Task} containing the next document if available or a failed task with
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoCursor.java:54]: This doesn't look right, it returns a ResultT?

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoIterable.java [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+package io.realm.mongodb.mongo.iterable;
+
+import com.google.android.gms.tasks.Task;
+
+import java.util.Collection;
+
+import io.realm.internal.common.TaskDispatcher;
+
+/**
+ * The MongoIterable is the results from an operation, such as a query.
```
Comment (cmelchior) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoIterable.java:26]: Good description 👍

File: realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java (hunk 4 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/internal/objectstore/OsMongoIterable.java:31]: Sure, I'll get it done.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoCursor.java (hunk 23 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoCursor.java:54]: Possible paste error when took the description from Stitch.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java (hunk 16 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoCollection.java:245]: It is the wrong description, it should specify that it uses options to build the query.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java (hunk 18 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java:41]: Becomes a private method after removing the unnecessary OS classes.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java (hunk 19 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/AggregateIterable.java:46]: Same as above.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/FindIterable.java (hunk 20 above)
Comment (edualonso) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/FindIterable.java:44]: Same as above.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoIterable.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (rorbech) [reply realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/iterable/MongoIterable.java:72]: Is it intended that calling this multiple times should trigger multiple remote operations? Maybe I am still a bit confused with the various _collection/iterable_ terms.

File: realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoClient.java [hunk 25]
```diff
@@ -18,21 +18,25 @@
...
 
-import io.realm.mongodb.User;
 import io.realm.internal.Util;
+import io.realm.internal.common.TaskDispatcher;
 import io.realm.internal.objectstore.OsMongoClient;
 
 /**
  * The remote MongoClient used for working with data in MongoDB remotely via Realm.
  */
 abstract public class MongoClient {
```
Comment (rorbech) [inline realm/realm-library/src/objectServer/java/io/realm/mongodb/mongo/MongoClient.java:28]: I recall adding getters for `App` and `User` on `Functions` for convenience. But can't recall the use case...maybe something like `functions.getApp().getConfiguration().getDefaultCodecRegistry()`. 

Maybe we should consider adding these for consistency. @cmelchior can you remember the use case?

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoClientTest.kt [hunk 26]
```diff
@@ -0,0 +1,1076 @@
...
+            // Getting a new iterator returns first element on tryNext
+            val asyncIter = iter.iterator().blockingGetResult()!!
+            assertEquals(doc1, asyncIter.tryNext().withoutId())
+        }
+    }
+
+    @Test
+    fun find_fails() {
+        with(getCollectionInternal()) {
+            assertFailsWith(ObjectServerError::class) {
```
Comment (rorbech) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoClientTest.kt:254]: There is already a `assertFailsWithErrorCode(expectedCode: ErrorCode, method: () -> Unit)` in `KotlinTestUtils.kt` maybe build on top of that. Could be changes in to a chain'able construct, so you could assert further details like you do.

File: realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoClientTest.kt [hunk 27]
```diff
@@ -0,0 +1,1076 @@
...
+            iter = aggregate(listOf(Document("\$match", doc1)))
+            assertTrue(iter.iterator().blockingGetResult()!!.hasNext())
+            assertEquals(doc1.withoutId(), iter.iterator().blockingGetResult()!!.next().withoutId())
+        }
+    }
+
+    @Test
+    fun aggregate_fails() {
+        with(getCollectionInternal()) {
+            assertFailsWith(ObjectServerError::class) {
```
Comment (rorbech) [inline realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/MongoClientTest.kt:295]: Same

**Pull request merged**